src/HappyEyeBallsConnectionBuilder.php

```diff
@@ -0,0 +1,304 @@
+                    $uri .= $that->parts['scheme'] . '://';
+                }
+
+                if (\strpos($ip, ':') !== false) {
+                    // enclose IPv6 addresses in square brackets before appending port
+                    $uri .= '[' . $ip . ']';
+                } else {
+                    $uri .= $ip;
+                }
+
+                // append original port if known
+                if (isset($that->parts['port'])) {
+                    $uri .= ':' . $that->parts['port'];
+                }
+
+                // append orignal path if known
+                if (isset($that->parts['path'])) {
+                    $uri .= $that->parts['path'];
+                }
+
+                // append original query if known
+                if (isset($that->parts['query'])) {
+                    $uri .= '?' . $that->parts['query'];
+                }
+
+                // append original hostname as query if resolved via DNS and if
+                // destination URI does not contain "hostname" query param already
+                $args = array();
+                \parse_str(isset($that->parts['query']) ? $that->parts['query'] : '', $args);
+                if ($that->host !== $ip && !isset($args['hostname'])) {
+                    $uri .= (isset($that->parts['query']) ? '&' : '?') . 'hostname=' . \rawurlencode($that->host);
+                }
+
+                // append original fragment if known
+                if (isset($that->parts['fragment'])) {
+                    $uri .= '#' . $that->parts['fragment'];
+                }
+
+                $promise = $that->connector->connect($uri);
+                $promise->then($resolve, $reject);
+            },
+            function ($_, $reject) use (&$promise, &$resolved, $that) {
+                // cancellation should reject connection attempt
+                // (try to) cancel pending connection attempt
+                if ($resolved === null) {
+                    $reject(new \RuntimeException('Connection to ' . $that->uri . ' cancelled during connection attempt'));
+                }
+
+                if ($promise instanceof CancellablePromiseInterface) {
+                    // overwrite callback arguments for PHP7+ only, so they do not show
+                    // up in the Exception trace and do not cause a possible cyclic reference.
+                    $_ = $reject = null;
+
+                    $promise->cancel();
+                    $promise = null;
+                }
+            }
+        );
+    }
+
+    /**
+     * @internal
+     */
+    public function cleanUp()
+    {
+        /** @var CancellablePromiseInterface $promise */
+        foreach ($this->connectionPromises as $index => $connectionPromise) {
+            if ($connectionPromise instanceof CancellablePromiseInterface) {
+                $connectionPromise->cancel();
+            }
+            unset($this->connectionPromises[$index]);
+            $connectionPromise = null;
```

Comment:
What's going on with these `null` assignments? (see also below)

Reply:
Hehehe was attempting to hun/prevent memory leaks :D

Reply:
Cleaned those out